src/storage/src/byte_array.cairo

```diff
@@ -0,0 +1,176 @@
+const BYTE_ARRAY_KEY: felt252 = '__BYTE_ARRAY__';
+const PENDING_WORD_KEY: felt252 = '__PENDING_WORD__';
+const PENDING_WORD_LEN_KEY: felt252 = '__PENDING_WORD_LEN__';
+
+/// Implementation of `LegacyHash` for `ByteArray`.
+/// The hash is given by the poseidon hash of the `data` and `pending_word` fields.
+impl LegacyHashByteArray of LegacyHash<ByteArray> {
+    fn hash(state: felt252, value: ByteArray) -> felt252 {
+        let mut buffer = array![value.pending_word];
+
+        let mut index = 0;
+
+        loop {
+            if index == value.data.len() {
+                break;
+            }
+
+            buffer.append((*value.data[index]).into());
+
+            index += 1;
+        };
+
+        poseidon::poseidon_hash_span(buffer.span())
+    }
+}
+
+/// Computes a storage address based on the given key.
+///
+/// # Arguments
+///
+/// * `base` - The base storage address (usually given by the name of the field in storage).
+/// * `key` - The custom key to add in a pre-defined array on which poseidon hash is computed.
+fn compute_storage_address(base: StorageBaseAddress, key: felt252) -> StorageBaseAddress {
+    let storage_addr = storage_address_to_felt252(storage_address_from_base(base));
+    let elt_keys: Array<felt252> = array![BYTE_ARRAY_KEY, storage_addr, key];
+
+    storage_base_address_from_felt252(poseidon::poseidon_hash_span(elt_keys.span()))
+}
+
+impl StoreByteArray of Store<ByteArray> {
+    fn read(address_domain: u32, base: StorageBaseAddress) -> SyscallResult<ByteArray> {
+        StoreByteArray::read_at_offset(address_domain, base, 0)
+    }
+
+    fn write(address_domain: u32, base: StorageBaseAddress, value: ByteArray) -> SyscallResult<()> {
+        StoreByteArray::write_at_offset(address_domain, base, 0, value)
+    }
+
+    fn read_at_offset(
+        address_domain: u32, base: StorageBaseAddress, offset: u8
+    ) -> SyscallResult<ByteArray> {
+        let data_len: usize = storage_read_syscall(
+            address_domain, storage_address_from_base_and_offset(base, offset)
+        )
+            .unwrap_syscall()
+            .try_into()
+            .expect('data len out of range');
+
+        let pw_base = compute_storage_address(base, PENDING_WORD_KEY);
+        let pending_word = storage_read_syscall(address_domain, storage_address_from_base(pw_base))
+            .unwrap_syscall();
+
+        let pwl_base = compute_storage_address(base, PENDING_WORD_LEN_KEY);
+        let pending_word_len: usize = storage_read_syscall(
+            address_domain, storage_address_from_base(pwl_base)
+        )
+            .unwrap_syscall()
+            .try_into()
+            .expect('pending_word_len out of range');
+
+        if data_len == 0 {
+            return Result::Ok(ByteArray { data: array![], pending_word, pending_word_len, });
+        }
+
+        let mut data: Array<bytes31> = array![];
+        let mut index = 0;
+
+        loop {
+            if index == data_len {
+                break ();
```

Comment:
```suggestion
                break;
```

Reply:
Fixed in 00e14928a0e72b1a8d58c89422fb7e18515c773f.